demo/lightning/checkpoint/multinode/fsspecfsdp.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+import gcsfs
+import torch
+
+from pathlib import Path
+from typing import Generator
+from lightning.pytorch.trainer.states import TrainerFn
+from lightning.pytorch.strategies import FSDPStrategy
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+from lightning.pytorch.strategies.fsdp import _METADATA_FILENAME
+from torch.distributed.checkpoint import save, load
+from torch.nn import Module
+from torch.distributed.checkpoint import _fsspec_filesystem as FF
+
+
+class FSSpecFSDPStrategy(FSDPStrategy):
```

Comment:
This class appears to have a lot of duplication with the `DatafluxFSDPStrategy`. I'd suggest thinking about what is different, and how you can abstract that away and have your FSSpecFSDPStrategy inherit DatafluxFSDPStrategy. Here's an (untested) example derived from some similar work I'm doing:

```python

class DatafluxFSDPStrategy(FSDPStrategy):

    def _init_io(self):
        self.writer = GCSDistributedWriter(path, project_name, storage_client)
        self.reader = GCSDistributedReader(path, project_name, storage_client)
        self.checkpoint_io = DatafluxLightningCheckpoint(
            project_name, storage_client)

    def _save(self, converted_state, path):
        save(converted_state, checkpoint_id=path, storage_writer=self.writer)

    def _write_checkpoint(self, checkpoint, path):
        self.checkpoint_io.save_checkpoint(checkpoint, path / _METADATA_FILENAME)

    def save_checkpoint(self, checkpoint, filepath, storage_options=None) -> None:
        if storage_options is not None:
            raise TypeError(...)

        path = Path(self.broadcast(filepath))

        converted_state = {"model": checkpoint.pop("state_dict")}
        converted_state.update({
            f"optimizer_{idx}": optim_state
            for idx, optim_state in enumerate(
                checkpoint.pop("optimizer_states", []))
        })
        self._save(converted_state, path)

        if self.global_rank == 0:
            self._write_checkpoint()


class FSSpecFSDPStrategy(DatafluxFSDPStrategy):

    def _init_io(self):
        self.writer = FF.FsspecWriter(self.path, sync_files=False)
        self.reader = FF.FsspecWriter(self.path, sync_files=False)

    def _write_checkpoint(self, path):
        bucket = gcsfs.GCSFileSystem()
        with bucket.open(os.path.join(filepath, _METADATA_FILENAME), 'wb') as f:
            torch.save(checkpoint, f)
```

This shows how we isolate the areas that change and move them into a private method that we can override with the changes specific to our different strategy. This also helps make the demo more readable as it highlights what needs to change from the default implementation.

Reply:
Or perhaps leave your code as is and we can refactor at a later time.

Reply:
This looks much better, Adam! You're right that `FSSpecFSDPStrategy` has a lot of duplication with the `DatafluxFSDPStrategy`. In fact, the former is just a copy-paste of the latter. I replaced Dataflux specific code with that of FSSpec. It took a lot of trial and error to figure out what changes to make and where.

I will file a follow-up task and work on making the changes you suggested here. Thanks for the example implementation!